webcrypto/hmac.go

```diff
@@ -0,0 +1,172 @@
+	// was set by the user, or if it was set to the default value.
+	lengthSet bool
+}
+
+// newHmacKeyGenParams creates a new HmacKeyGenParams object, from the normalized
+// algorithm, and the params parameters passed by the user.
+//
+// It handles the logic of extracting the hash algorithm from the params object,
+// and normalizing it. It also handles the logic of extracting the length
+// attribute from the params object, and setting it to the default value if it's
+// not present as described in the hmac `generateKey` [specification].
+//
+// [specification]: https://www.w3.org/TR/WebCryptoAPI/#hmac-operations
+func newHmacKeyGenParams(rt *goja.Runtime, normalized Algorithm, params goja.Value) (*HmacKeyGenParams, error) {
+	// The specification doesn't explicitly tell us what to do if the
+	// hash field is not present, but we assume it's a mandatory field
+	// and throw an error if it's not present.
+	hashValue, err := traverseObject(rt, params, "hash")
+	if err != nil {
+		return nil, NewError(0, SyntaxError, "could not get hash from algorithm parameter")
+	}
+
+	// Although the specification doesn't explicitly ask us to do so, we
+	// normalize the hash algorithm here, as it shares the same definition
+	// as the AlgorithmIdentifier type, and we'll need it later on.
+	//
+	// Note @oleiade: The specification seems to assume that the normalization
+	// algorithm will normalize the existing Algorithm fields, and leave
+	// the rest untouched. As we are in the context of a statically typed
+	// language, we can't do that, so we need to normalize the hash
+	// algorithm here.
+	normalizedHash, err := normalizeAlgorithm(rt, hashValue, OperationIdentifierGenerateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// As the length attribute is optional and as the key generation process
+	// differentiates unset from zero-values, we need to handle the case
+	// where the length attribute is not present in the params object.
+	var length int
+	var lengthSet bool
+	algorithmLengthValue, err := traverseObject(rt, params, "length")
+	if err == nil {
+		lengthSet = true
+		length = int(algorithmLengthValue.ToInteger())
+	}
+
+	return &HmacKeyGenParams{
+		Algorithm: normalized,
+		Hash:      normalizedHash,
+		Length:    length,
+
+		lengthSet: lengthSet,
+	}, nil
+}
+
+// GenerateKey generates a new HMAC key.
+func (hkgp *HmacKeyGenParams) GenerateKey(
+	extractable bool,
+	keyUsages []CryptoKeyUsage,
+) (*CryptoKey, error) {
+	// 1.
+	for _, usage := range keyUsages {
+		switch usage {
+		case SignCryptoKeyUsage, VerifyCryptoKeyUsage:
+			continue
+		default:
+			return nil, NewError(0, SyntaxError, "invalid key usage: "+usage)
+		}
+	}
+
+	// 2.
+	// We extract the length attribute from the algorithm object, as it's not
+	// part of the normalized algorithm, and as accessing the runtime from the
+	// callback below could lead to a race condition.
+	if !hkgp.lengthSet {
+		switch hkgp.Hash.Name {
+		case Sha1:
+			hkgp.Length = sha1.BlockSize * 8
```

Comment:
You're completely right, I'm not happy with it either. I'm thinking we could have a type `Bit`, and a type `Byte` with set values, that one can multiply against, in the same fashion as with `time.Duration`. What do you think? Otherwise conversion methods like you proposed would work too 👍🏻

Reply:
I feel like that a `type` is that you won't be able to just use as an `int` directly, but calling `BitsLength()` will return you an int which you can then use.

Having something like time.Second and co, will in practice means that you will need to everywhere cast it to `int`. Or you will jsut need methods on it as well.

But let's make an issue out of this?

Reply:
https://github.com/grafana/xk6-webcrypto/issues/36#issue-1647527795